contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol

```diff
@@ -0,0 +1,582 @@
+import { IController } from "../interfaces/IController.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { IWETH } from "../interfaces/IWETH.sol";
+import { PreciseUnitMath } from "../lib/PreciseUnitMath.sol";
+
+
+contract ExchangeIssuanceZeroEx is Ownable, ReentrancyGuard {
+
+    using Address for address payable;
+    using SafeMath for uint256;
+    using PreciseUnitMath for uint256;
+    using SafeERC20 for IERC20;
+    using SafeERC20 for ISetToken;
+
+    struct ZeroExSwapQuote {
+        IERC20 sellToken;
+        IERC20 buyToken;
+        bytes swapCallData;
+    }
+
+    struct IssuanceModuleData {
+        bool isAllowed;
+        bool isDebtIssuanceModule;
+    }
+
+    /* ============ Constants ============== */
+
+    address constant public ETH_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+    /* ============ State Variables ============ */
+
+    address public immutable WETH;
+    mapping(address => IssuanceModuleData) public allowedIssuanceModules;
+
+    IController public immutable setController;
+
+    address public swapTarget;
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(
+        address indexed _recipient,     // The recipient address of the issued SetTokens
+        ISetToken indexed _setToken,    // The issued SetToken
+        IERC20 indexed _inputToken,     // The address of the input asset(ERC20/ETH) used to issue the SetTokens
+        uint256 _amountInputToken,      // The amount of input tokens used for issuance
+        uint256 _amountSetIssued        // The amount of SetTokens received by the recipient
+    );
+
+    event ExchangeRedeem(
+        address indexed _recipient,     // The recipient address which redeemed the SetTokens
+        ISetToken indexed _setToken,    // The redeemed SetToken
+        IERC20 indexed _outputToken,    // The address of output asset(ERC20/ETH) received by the recipient
+        uint256 _amountSetRedeemed,     // The amount of SetTokens redeemed for output tokens
+        uint256 _amountOutputToken      // The amount of output tokens received by the recipient
+    );
+
+    event Refund(
+        address indexed _recipient,     // The recipient address which redeemed the SetTokens
+        uint256 _refundAmount           // The amount of ETH redunded to the recipient
+    );
+
+    event BoughtTokens(IERC20 sellToken, IERC20 buyToken, uint256 boughtAmount);
+
+    /* ============ Modifiers ============ */
+
+    modifier isSetToken(ISetToken _setToken) {
+         require(setController.isSet(address(_setToken)), "ExchangeIssuance: INVALID SET");
+         _;
+    }
+
+    modifier isWhitelistedIssuanceModule(address _issuanceModule) {
+        require(allowedIssuanceModules[_issuanceModule].isAllowed, "ExchangeIssuance: INVALID ISSUANCE MODULE");
+         _;
+    }
+
+    modifier isValidInput(ISetToken _setToken, uint256 _amountSetToken, ZeroExSwapQuote[] memory _componentQuotes) {
```

Comment:
What happens if a user passes in malformed data here? Can they use it to attack the contract? Will they lose their funds? If we pass in zero for the amount, it will revert on the issuance call (since both issuance modules already check if the amount is 0). If we pass in bad swaps, it will either fail when it tries to issue (and it hasn't bought the right components), or it will fail when we do the slippage check for redemption (since we likely would have forgotten to sell a component).

I think we should remove this. We can ensure all of these things at the frontend.

Reply:
True, I don't think it is security relevant. Might save people some gas by reverting earlier. But since these conditions are not dependant on contract state they should come up during "gasEstimation" already anyway, before a transaction is submitted.

Will remove it as suggested